src/Algebra/Graph/LabelledAdjacencyMap.hs

```diff
@@ -0,0 +1,780 @@
+-- 'isEmpty' . connects == 'all' 'isEmpty'
+-- @
+connects
+        :: (Ord a, Dioid e)
+        => [LabelledAdjacencyMap a e]
+        -> LabelledAdjacencyMap a e
+connects = foldr connect empty
+
+-- | The 'isSubgraphOf' function takes two graphs and returns 'True' if the
+-- first graph is a /subgraph/ of the second. Complexity: /O((n + m) * log(n))/
+-- time.
+--
+-- @
+-- isSubgraphOf 'empty'         x             == True
+-- isSubgraphOf ('vertex' x)    'empty'         == False
+-- isSubgraphOf x             ('overlay' x y) == True
+-- isSubgraphOf ('overlay' x y) ('connect' x y) == True
+-- isSubgraphOf ('path' xs)     ('circuit' xs)  == True
+-- @
+isSubgraphOf
+        :: (Ord a, Eq e)
+        => LabelledAdjacencyMap a e
+        -> LabelledAdjacencyMap a e
+        -> Bool
+isSubgraphOf x y = Map.isSubmapOfBy Map.isSubmapOf
+                                    (labelledAdjacencyMap x)
+                                    (labelledAdjacencyMap y)
+
+-- | Check if a graph is empty.
+-- Complexity: /O(1)/ time.
+--
+-- @
+-- isEmpty 'empty'                       == True
+-- isEmpty ('overlay' 'empty' 'empty')       == True
+-- isEmpty ('vertex' x)                  == False
+-- isEmpty ('removeVertex' x $ 'vertex' x) == True
+-- isEmpty ('removeEdge' x y $ 'edge' x y) == False
+-- @
+isEmpty :: LabelledAdjacencyMap a e -> Bool
+isEmpty = Map.null . labelledAdjacencyMap
+
+-- | Check if a graph contains a given vertex.
+-- Complexity: /O(log(n))/ time.
+--
+-- @
+-- hasVertex x 'empty'            == False
+-- hasVertex x ('vertex' x)       == True
+-- hasVertex 1 ('vertex' 2)       == False
+-- hasVertex x . 'removeVertex' x == const False
+-- @
+hasVertex :: Ord a => a -> LabelledAdjacencyMap a e -> Bool
+hasVertex x = Map.member x . labelledAdjacencyMap
+
+-- | Check if a graph contains a given edge.
+-- Complexity: /O(log(n))/ time.
+--
+-- @
+-- hasEdge x y 'empty'            == False
+-- hasEdge x y ('vertex' z)       == False
+-- hasEdge x y ('edge' x y)       == True
+-- hasEdge x y . 'removeEdge' x y == const False
+-- hasEdge x y                  == 'elem' (x,y) . 'edgeList'
+-- @
+hasEdge :: Ord a => a -> a -> LabelledAdjacencyMap a e -> Bool
+hasEdge u v a = case Map.lookup u (labelledAdjacencyMap a) of
+        Nothing -> False
+        Just vs -> Map.member v vs
+
+
+-- | The number of vertices in a graph.
+-- Complexity: /O(1)/ time.
+--
+-- @
+-- vertexCount 'empty'      == 0
+-- vertexCount ('vertex' x) == 1
+-- vertexCount            == 'length' . 'vertexList'
+-- @
+vertexCount :: LabelledAdjacencyMap a e -> Int
+vertexCount = Map.size . labelledAdjacencyMap
+
+-- | The number of edges in a graph.
+-- Complexity: /O(n)/ time.
+--
+-- @
+-- edgeCount 'empty'      == 0
+-- edgeCount ('vertex' x) == 0
+-- edgeCount ('edge' x y) == 1
+-- edgeCount            == 'length' . 'edgeList'
+-- @
+edgeCount :: LabelledAdjacencyMap a e -> Int
+edgeCount = Map.foldr (\es r -> (Map.size es + r)) 0 . labelledAdjacencyMap
```

Comment:
Note that Alga HEAD has a different implementation of this function for `AdjacencyMap`: 

```
edgeCount = getSum . foldMap (Sum . Set.size) . adjacencyMap
```